### PR TITLE
[MDS] Add Vega support for importing saved objects

### LINKED ISSUE #5927: [MDS] Vega support with MDS
## Proposal
Currently, Multiple DataSource (MDS) does not support Vega visualizations. Thus, we propose to create a new optional field from within the Vega-spec `data_source_name` (tentatively named) that is passed within the `url` body of the `data` body. This field will take in a datasource name and under the hood, the visualization will be able to retrieve data from the index located in that datasource. This enables a user to retrieve data from one or more indices from one or more datasources to create custom Vega visualizations.

Here is an example Vega spec that should be supported with this feature. Note that `data` field can be a singular `url` object or an array that contains multiple `url` objects:
```
data: {
    url: {
      %context%: true
      %timefield%: @timestamp
      index: opensearch_dashboards_sample_data_logs
      datasource: some_datasource_id
      body: {
        aggs: {
          ...
      }
    }
    format: {
      property: aggregations.5
    }
  }
```

## Background
[Vega](https://vega.github.io/vega/) is a declarative visualization grammar that can be used to create and share custom interactive visualizations. It is important to note that [Vega-lite](https://vega.github.io/vega-lite/) is a similar but very much different lightweight visualization grammar. Both of these grammars are supported within Dashboards but with the caveat that the data is retrieved from the index BEFORE any rendering happens. This means that data is NOT dynamically loaded. Additionally, Vega is not supported with MDS since the local cluster is the assumed datasource. This proposal will fix the latter (with the former being out of scope).

It is also important to clarify what is meant by Vega support for MDS. Vega support for MDS can be interpreted in two different ways
1. **Option 1:** Vega can support visualizations which fetch data from multiple datasources
MDS drawio
    In this example, the Vega visualization is stored in Datasource A but references data from Datasources A, B, and C

2. **Option 2:** Vega can support visualizations which reference data from the local cluster or any remote datasource (but not both) 
MDS drawio(1)
    In this example, the Vega visualization is stored in Datasource A but references data only from Datasource B and not any other datasource

The proposal seeks to support **Option 1**. While there are more limitations with option 1 (see the `Limitations` section), having the option to fetch data from any index from any datasource (provided the user has the permissions) provides a robust visualization experience.

## Approach
When Dashboards parses the Vega-spec to render the visualization, it [parses the URL object](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_type_vega/public/data_model/opensearch_query_parser.ts) and passes the object into the [search API](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_type_vega/public/data_model/search_api.ts), which uses `IOpenSearchSearchRequest` as a parameter. This interface provides a field `dataSourceId` that will tell dashboards to use the [data source client](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data_source/README.md). All that the Vega plugin would need to do is check if MDS is enabled and if so, retrieve the associated datasource id from the `data_source_name` field and pass it into the search query.

Add a `datasource` field to the `UrlObject`
```Typescript
export interface UrlObject {
  [index: string]: any;
  [CONSTANTS.TIMEFILTER]?: string;
  [CONSTANTS.CONTEXT]?: boolean;
  [CONSTANTS.LEGACY_CONTEXT]?: string;
  [CONSTANTS.TYPE]?: string;
  name?: string;
  index?: string;
  data_source_name?: string;
  body?: Body;
  size?: number;
  timeout?: string;
}
```

Have the saved objects client get the associated `dataSourceId` from the `data_source_name` via `find`

Then, in the `_searchAPI.search()` method, we can pass in the `dataSourceId` as a parameter
```Typescript
return search({params, dataSourceId} , { abortSignal: this.abortSignal }).pipe(
	tap((data) => this.inspectSearchResult(data, requestResponders[requestId])),
	map((data) => ({
		name: requestId,
		rawResponse: data.rawResponse,
	}))
);
```

Thus, when the user wants to write a visualization with data from another datasource, they can do something like the following (mockup)
image

## Limitations
This approach provides greater flexibility in enabling users to make visualizations. However with great power comes great responsibility. 
1. Importing/exporting Vega saved objects will inevitably break. Because `data_source_name` is specific to an OpenSearch cluster, if users were to export/import into another cluster and the same datasource names are not configured, the visualizations cannot find the data and thus return errors. This is a tradeoff that can be made but in the future, this would ideally need to be mitigated.
2. As a sort of add-on to point 1, editing `data_source_names` to each `url` that contains one is cumbersome, especially when multiple `url` objects were involved.

## Alternatives
Initially the decision was made to use the `data_source_id` vs the `data_source_name`. This was due to the fact that `data_source_id` enforces a unique datasource to query from and does not make an extra find query to find the datasource. However, other plugins referred to datasource by name, not by id, and having name be the identifier is more user friendly.

## Open Question(s)
1. Since `datasource` can be a bit ambiguous here, what are some alternative field names here? Im thinking `data_source_id` would help disambiguate this field

### PR DESCRIPTION
**Pull request by huyaboo**
### Description
This PR adds (limited) support for importing Vega saved objects when `data_source.enabled: true` and when specifying a datasource. Support in this case means:
- A user can import Vega Visualizations from non-MDS enabled clusters to MDS enabled clusters in `create` mode or in `overwrite` mode
- A user can import Vega Visualizations from other MDS enabled clusters to MDS enabled clusters in `create` mode or in `overwrite` mode

Limitations:
- Only one `data_source_name` can be changed; if users have multiple `data_source_names` specified in their Vega spec, the origin datasource name will be changed but the others will stay the same
- Possibility of duplicate Data-source saved objects being created. The user may potentially have to handle that

### Issues Resolved

Closes #5927 

## Screenshot

When updating the `data_source_name`s, the datasource references also change. In this recording, `Data Source A` is updated to `Data Source B`

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/73027756/bda9745e-8cea-4ad9-b132-4577938eb1d8

Importing a similar visualization that references `Data Source B` and has a local query will update the spec to include `Data Source A` in the old local queries. `Data Source A` is also shown to be added as a reference

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/73027756/ba050fc1-61ab-46c7-bbb2-68d68636c983

https://github.com/opensearch-project/OpenSearch-Dashboards/assets/73027756/58f006d4-e076-47bd-b519-291e3fcac160

## Testing the changes
The following scenarios were tested:

- [x] Import from non-MDS cluster -> non-MDS cluster (existing functionality should not be impacted)
- [x] Import from non-MDS cluster -> MDS cluster (new datasource name should be attached to each OpenSearch query)
- [x] Import from MDS cluster -> MDS cluster (all local queries will be replaced with the new `data_source_name` and the references will append this new datasource id)

### Check List

- [x] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/core/server/saved_objects/import/check_conflict_for_data_source.ts [hunk 1]
```diff
@@ -74,12 +92,37 @@ export async function checkConflictsForDataSource({
...
          */
+
+        // Some visualization types will need special modifications, like Vega visualizations
+        if (object.type === 'visualization') {
+          const vegaSpec = extractVegaSpecFromSavedObject(object);
+
+          if (!!vegaSpec) {
+            const previousDataSourceTitle =
+              previoudDataSourceId && savedObjectsClient
+                ? await getDataSourceTitleFromId(previoudDataSourceId, savedObjectsClient)
```
Comment (BionIT) [inline src/core/server/saved_objects/import/check_conflict_for_data_source.ts:103]: in this logic, previousDataSourceId is something we get from object id which is not the case for vega right, since we don't save the data source id into vega vis object id?

File: src/core/server/saved_objects/import/check_conflict_for_data_source.ts [hunk 2]
```diff
@@ -74,12 +92,37 @@ export async function checkConflictsForDataSource({
...
+                : undefined;
```
Comment (BionIT) [inline src/core/server/saved_objects/import/check_conflict_for_data_source.ts:104]: Can we avoid making individual calls if we can make a bulk find request?

File: src/core/server/saved_objects/import/create_saved_objects.ts [hunk 3]
```diff
@@ -83,88 +88,117 @@ export const createSavedObjects = async <T>({
...
-        // @ts-expect-error
-        const searchSourceString = object.attributes?.kibanaSavedObjectMeta?.searchSourceJSON;
-        // @ts-expect-error
-        const visStateString = object.attributes?.visState;
+        // Some visualization types will need special modifications, like Vega visualizations
+        if (object.type === 'visualization') {
+          const vegaSpec = extractVegaSpecFromSavedObject(object);
+
+          if (!!vegaSpec && !!dataSourceTitle) {
+            const idComponents = object.id.split('_');
```
Comment (BionIT) [inline src/core/server/saved_objects/import/create_saved_objects.ts:111]: Do we save the data source id as part of vega vis object id?

**Comment by BionIT** [comment]
Can we add the manual test details for this PR?

File: src/core/server/saved_objects/import/utils.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    if (jsonSpec && typeof jsonSpec === 'object') {
+      return jsonSpec;
+    }
+  } catch (e) {
+    return false;
+  }
+
+  return false;
+};
```
Comment (xinruiba) [inline src/core/server/saved_objects/import/utils.ts:107]: Instead of returning a json when condition meets and return a boolean when failed.
What if directly return undefined when condition not meet?

```
const parseJSONSpec = (spec: string) => {
  try {
    const jsonSpec = JSON.parse(spec);

    if (jsonSpec && typeof jsonSpec === 'object') {
      return jsonSpec;
    }
  } catch (e) {
    return;
  }

  return;
};
```

File: src/core/server/saved_objects/import/utils.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+export const extractVegaSpecFromSavedObject = (savedObject: SavedObject) => {
+  if (isVegaVisualization(savedObject)) {
+    // @ts-expect-error
+    const visStateObject = JSON.parse(savedObject.attributes?.visState);
+    return visStateObject.params.spec;
+  }
+
+  return false;
```
Comment (xinruiba) [inline src/core/server/saved_objects/import/utils.ts:62]: Nit:
Since the function named as `extractVegaSpecFromSavedObject`, the called of this func will be expect an object.
What if we return undefined here instead of a boolean. What do you think?

File: src/core/server/saved_objects/import/check_conflict_for_data_source.ts [hunk 6]
```diff
@@ -43,7 +55,13 @@ export async function checkConflictsForDataSource({
...
     new Map<string, SavedObjectsImportRetry>()
   );
-  objects.forEach((object) => {
+
+  const dataSourceTitle =
+    !!dataSourceId && !!savedObjectsClient
+      ? await getDataSourceTitleFromId(dataSourceId, savedObjectsClient)
+      : undefined;
+
+  for await (const object of objects) {
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.ts]: No I have since made the loop synchronous

File: src/core/server/saved_objects/import/check_conflict_for_data_source.ts (hunk 1 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.ts]: Yes. Now since we attach datasource references, we do not need a previous datasource id. When a vega saved object is imported from to a datasource, any local queries will be modified to append a `data_source_name` and the new datasourceid reference attached to the references list.

File: src/core/server/saved_objects/import/check_conflict_for_data_source.ts (hunk 2 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.ts]: Removed repeated individual `get` requests

File: src/core/server/saved_objects/import/create_saved_objects.ts (hunk 3 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/create_saved_objects.ts]: Yes. The new datasourceid is still part of the vega vis object id and the reference to that datasourceid is also created.

File: src/core/server/saved_objects/import/utils.ts (hunk 5 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/utils.ts]: Changed to `undefined`

File: src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  findDataSourceIdbyName,
+} from './utils';
+import { getDataSourceEnabled } from './services';
+
+export const VEGA_VISUALIZATION_CLIENT_WRAPPER_ID = 'vega-visualization-client-wrapper';
+
+export const vegaVisualizationClientWrapper: SavedObjectsClientWrapperFactory = (
+  wrapperOptions: SavedObjectsClientWrapperOptions
+) => {
+  const createForVega = async <T = unknown>(
```
Comment (BionIT) [inline src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts:24]: for create, I think there are a few validations we can do here when mds is enabled:
1. there must be reference data source fields and data source name fields in the data object
2. the data source references must exist - using bulk get
3. the data source references names fetched from 2 should match with all data source names

that means when we call this create API, we need to be able to get the data source references

File: src/plugins/vis_type_vega/server/utils.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  if (possibleDataSourceIds.length !== 1) {
+    throw new Error(
+      `Expected exactly 1 result for data_source_name "${dataSourceName}" but got ${possibleDataSourceIds.length} results`
+    );
+  }
+
+  return possibleDataSourceIds.pop()?.id;
+};
+
+export const extractVegaSpecFromAttributes = (attributes: unknown) => {
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/utils.ts:33]: Visualization types (Vega, area, timeline) are determined by parsing the stringified `visState` of the object, which is in the `attributes` of a saved object. What's being passed here is the `attributes` field of a visualization saved object, which may or may not be a Vega visualization.

File: src/plugins/vis_type_vega/server/utils.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+      dataSourceNameSet.add(dataSourceName);
+    }
+  } else {
+    throw new Error(`"data" field should be an object or an array of objects`);
+  }
+
+  return dataSourceNameSet;
+};
+
+const getDataSourceNameFromObject = (dataObject: any) => {
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/utils.ts:67]: `dataObject` is parsed from stringified JSON/HJSON, which doesn't have a type at compile time. Additionally, there are multiple types of `dataObjects` as per the Vega [Data](https://vega.github.io/vega/docs/data/) field.

File: src/plugins/vis_type_vega/server/plugin.ts [hunk 10]
```diff
@@ -44,10 +49,19 @@ export class VisTypeVegaPlugin implements Plugin<VisTypeVegaPluginSetup, VisType
...
+  public setup(
+    core: CoreSetup,
+    { home, usageCollection, dataSource }: VisTypeVegaPluginSetupDependencies
+  ) {
     if (usageCollection) {
       registerVegaUsageCollector(usageCollection, this.config, { home });
     }
+    setDataSourceEnabled({ enabled: dataSource?.dataSourceEnabled() || false });
+    core.savedObjects.addClientWrapper(
+      10,
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/plugin.ts:61]: `10` is the priority number (required) for adding client wrapper. This may need setting up

File: src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts (hunk 7 above)
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts:24]: 1. The only way for a Vega spec to be updated through UI is from the save button. The save button, whether overwriting the existing saved object or creating a new visualization, will call `create` to the saved objects client. Thus, from UI perspective, a saved object cannot have a `data_source_name` field but no associated references. The only way a saved object does not have references is if the Vega spec does not specify a `data_source_name` (local cluster queries only).
2. If a non-existent datasource reference (like a datasource that has been deleted) will be filtered out when updating the `references` field because calling bulk get on the object will return the `id` field and an `error` field instead of the `attributes` field.
```Typescript
(await wrapperOptions.client.bulkGet(existingDataSourceReferences)).saved_objects.forEach(
  (object) => {
    // @ts-expect-error
    if (!!object.attributes && !!object.attributes.title) {
      // @ts-expect-error
      existingDataSourceIdToNameMap.set(object.id, object.attributes.title);
    }
  }
);
```
3. This has also been checked. The datasource reference name must be in the `dataSourceNamesSet` generated from the Vega spec.
```Typescript
const newReferences = options?.references?.filter((reference) => {
  if (reference.type !== 'data-source') {
    return true;
  }
  const dataSourceName = existingDataSourceIdToNameMap.get(reference.id);
  if (dataSourceNamesSet.has(dataSourceName)) {
    dataSourceNamesSet.delete(dataSourceName);
    return true;
  }

  return false;
});
```

**Note:** 
Before saving a Vega visualization, the user must press `update` on the visualization renderer. If a user were to type a nonexistent/outdated `data_source_name`, the renderer should catch the error and return the message to the user before saving. This means that a non-existent datasource name cannot be in the Vega spec (unless the user ignores the error message and saves anyway).

There is a chance that the user will edit the data source names in the data sources page but not the Vega spec and export, which will cause a mismatch in the `data_source_name` and the actual reference. However, this edge case is hard to check because if the visualization has more than one data source reference and the names of these references were all changed, we cannot map the old names -> updated names.

File: src/plugins/vis_type_vega/server/utils.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  dataSourceName: string;
+  savedObjectsClient: SavedObjectsClientContract;
+}
+
+export const findDataSourceIdbyName = async (props: FindDataSourceByTitleQueryProps) => {
+  const { dataSourceName } = props;
+  const dataSources = await dataSourceFindQuery(props);
+
+  // In the case that data_source_name is a prefix of another name, match exact data_source_name
+  const possibleDataSourceIds = dataSources.saved_objects.filter(
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/utils.ts]: changed to `possibleDataSourceObjects`

File: src/core/server/saved_objects/import/utils.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+export interface UpdateDataSourceNameInVegaSpecProps {
+  spec: string;
+  newDataSourceName: string;
+}
+
+export const updateDataSourceNameInVegaSpec = (
+  props: UpdateDataSourceNameInVegaSpecProps
+): string => {
+  // Exit early if import data cannot determine new datasource name
+  if (!!props.newDataSourceName === false) {
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/core/server/saved_objects/import/utils.ts]: Removed the check since all calls to this function checks if `dataSourceTitle` exists.

File: src/core/server/saved_objects/import/utils.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        keepWsc: true,
+      });
+};
+
+export const getDataSourceTitleFromId = async (
+  dataSourceId: string,
+  savedObjectsClient: SavedObjectsClientContract
+) => {
+  return await savedObjectsClient.get('data-source', dataSourceId).then((response) => {
+    const attributes: any = response?.attributes || undefined;
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/core/server/saved_objects/import/utils.ts]: Refactored

File: src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    attributes: T,
+    options?: SavedObjectsCreateOptions
+  ) => {
+    const vegaSpec = extractVegaSpecFromAttributes(attributes);
+    if (type !== 'visualization' || vegaSpec === undefined || !getDataSourceEnabled().enabled) {
+      return await wrapperOptions.client.create(type, attributes, options);
+    }
+    const dataSourceNamesSet = extractDataSourceNamesInVegaSpec(vegaSpec);
+
+    const existingDataSourceIds = options?.references
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts]: Changed to `existingDataSourceReferences`

File: src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+      });
+
+    const existingDataSourceNamesMap = new Map();
+    if (!!existingDataSourceIds && existingDataSourceIds.length > 0) {
+      (await wrapperOptions.client.bulkGet(existingDataSourceIds)).saved_objects.forEach(
+        (object) => {
+          // @ts-expect-error
+          if (!!object.attributes && !!object.attributes.title) {
+            // @ts-expect-error
+            existingDataSourceNamesMap.set(object.id, object.attributes.title);
```
[the remark this answers is not in the document]
Comment (huyaboo) [reply src/plugins/vis_type_vega/server/vega_visualization_client_wrapper.ts]: Changed to `existingDataSourceIdToNameMap`

File: src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts [hunk 16]
```diff
@@ -140,4 +180,82 @@ describe('#checkConflictsForDataSource', () => {
...
+        errors: [],
+        importIdMap: new Map([
+          [
+            `visualization:some-object-id`,
+            { id: 'some-datasource-id_some-object-id', omitOriginId: true },
+          ],
+        ]),
+      })
+    );
+  });
```
Comment (bandinib-amzn) [inline src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts:226]: minor: for better coverage, maybe we want to test that savedobjectclient.get have been called? And also test called with expected attributes?

File: src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts [hunk 17]
```diff
@@ -140,4 +180,82 @@ describe('#checkConflictsForDataSource', () => {
...
```
Comment (bandinib-amzn) [inline src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts:260]: same as above

Comment (bandinib-amzn) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts:260]: also can we add test for duplicate and undefined data source title?

**Review by BionIT — APPROVED** [review]

File: src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts (hunk 16 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts:230]: Done, added a test to ensure `get()` was called.

File: src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts (hunk 17 above)
Comment (huyaboo) [reply src/core/server/saved_objects/import/check_conflict_for_data_source.test.ts:268]: Since this makes a `get` by id, I don't think we need to check for duplicate titles here since there is no ambiguity. The duplicate title check will be done when the visualization is saved and when the visualization is rendered.

**Review by BionIT — APPROVED** [review]

**Review by ZilongX — DISMISSED** [review]

**Review by BionIT — APPROVED** [review]

**Review by BionIT — APPROVED** [review]

**Review by bandinib-amzn — APPROVED** [review]
